src/txrequest.cpp

```diff
@@ -0,0 +1,724 @@
+        }
+        Modify<ByTxHash>(it, [new_state](Announcement& ann){ ann.m_state = new_state; });
+    }
+
+    //! Check if 'it' is the only announcement for a given txhash that isn't COMPLETED.
+    bool IsOnlyNonCompleted(Iter<ByTxHash> it)
+    {
+        assert(it != m_index.get<ByTxHash>().end());
+        assert(it->m_state != State::COMPLETED); // Not allowed to call this on COMPLETED announcements.
+
+        // This announcement has a predecessor that belongs to the same txhash. Due to ordering, and the
+        // fact that 'it' is not COMPLETED, its predecessor cannot be COMPLETED here.
+        if (it != m_index.get<ByTxHash>().begin() && std::prev(it)->m_txhash == it->m_txhash) return false;
+
+        // This announcement has predecessor that belongs to the same txhash, and is not COMPLETED.
+        if (std::next(it) != m_index.get<ByTxHash>().end() && std::next(it)->m_txhash == it->m_txhash &&
+            std::next(it)->m_state != State::COMPLETED) return false;
+
+        return true;
+    }
+
+    /** Convert any announcement to a COMPLETED one. If there are no non-COMPLETED announcements left for this
+     *  txhash, they are deleted. If this was a REQUESTED announcement, and there are other CANDIDATEs left, the
+     *  best one is made CANDIDATE_BEST. Returns whether the announcement still exists. */
+    bool MakeCompleted(Iter<ByTxHash> it)
+    {
+        assert(it != m_index.get<ByTxHash>().end());
+
+        // Nothing to be done if it's already COMPLETED.
+        if (it->m_state == State::COMPLETED) return true;
+
+        if (IsOnlyNonCompleted(it)) {
+            // This is the last non-COMPLETED announcement for this txhash. Delete all.
+            uint256 txhash = it->m_txhash;
+            do {
+                it = Erase<ByTxHash>(it);
+            } while (it != m_index.get<ByTxHash>().end() && it->m_txhash == txhash);
+            return false;
+        }
+
+        // Mark the announcement COMPLETED, and select the next best announcement (the first CANDIDATE_READY) if
+        // needed.
+        ChangeAndReselect(it, State::COMPLETED);
+
+        return true;
+    }
+
+    //! Make the data structure consistent with a given point in time:
+    //! - REQUESTED annoucements with expiry <= now are turned into COMPLETED.
+    //! - CANDIDATE_DELAYED announcements with reqtime <= now are turned into CANDIDATE_{READY,BEST}.
+    //! - CANDIDATE_{READY,BEST} announcements with reqtime > now are turned into CANDIDATE_DELAYED.
+    void SetTimePoint(std::chrono::microseconds now)
+    {
+        // Iterate over all CANDIDATE_DELAYED and REQUESTED from old to new, as long as they're in the past,
+        // and convert them to CANDIDATE_READY and COMPLETED respectively.
+        while (!m_index.empty()) {
+            auto it = m_index.get<ByTime>().begin();
+            if (it->m_state == State::CANDIDATE_DELAYED && it->m_time <= now) {
+                PromoteCandidateReady(m_index.project<ByTxHash>(it));
```

Comment:
Yes, that's expected. The loop is expected to make progress in every step, or break. With the change you're making, it will infinitely spin when reaching a CANDIDATE_DELAYED.